transient/sshfs.py

```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
-    is_slow: bool = False,
 ) -> None:
 
+    sshfs_options = "-o slave,allow_other"
+    sshfs_command = f"sudo -E sshfs {sshfs_options} :{local_dir} {remote_dir}"
+
+    # Because sshfs monopolizes stdout, progress markers go to stderr
+    ssh_command = f"""
+        set -e
+        echo TRANSIENT_SSH_COMPLETE >&2
+        sudo mkdir -p {remote_dir}
+        {sshfs_command}
+    """
+
     sshfs_config = ssh_config.override(
-        args=["-A", "-T", "-o", "LogLevel=ERROR",] + ssh_config.args
+        args=["-T", "-o", "LogLevel=ERROR"] + ssh_config.args
+    )
+    client = ssh.SshClient(sshfs_config, command=ssh_command)
+
+    logging.info(f"Sending sshfs mount command '{sshfs_command}'")
+    ssh_proc = client.connect(
+        timeout=connect_timeout,
+        stdin=sftp_proc.stdout,
+        stdout=sftp_proc.stdin,
+        stderr=subprocess.PIPE,
+        tie_to_thread=False,
     )
-    client = ssh.SshClient(sshfs_config)
-    conn = client.connect_piped(timeout=connect_timeout)
-
-    try:
-        sshfs_options = "-o StrictHostKeyChecking=no -o allow_other"
-        sshfs_command = f"sudo -E sshfs {sshfs_options} {local_user}@10.0.2.2:{local_dir} {remote_dir}"
-
-        logging.info(f"Sending sshfs mount command '{sshfs_command}'")
-
-        sshfs_timeout = _SSHFS_MAX_RUN_TIME
-        if is_slow is True:
-            sshfs_timeout = _SSHFS_MAX_RUN_TIME_SLOW
-
-        # This is somewhat gnarly. The core of the issue is that sshfs is a FUSE mount,
-        # so it runs as a process (that gets backgrounded by default). SSH won't close
-        # the connection on its side until "it encounters end-of-file (eof) on the pipes
-        # connecting  to the stdout and stderr of the user program". This typically means
-        # you can do something like 'nohup <cmd> >/dev/null </dev/null 2>&1 &' to close
-        # all handles and ignore any hang ups. However, this doesn't work for SSHFS, as
-        # it spawns other processes that (I guess?) still have an open handle.
-        #
-        # This causes the SSH connetion to hang forever after the logout. Therefore, we
-        # need to close the connection on our end. So the trick here is to wait for some
-        # max time for this process to be done, then inspect the stdout for a sentinel
-        # value indicating that we _did_ get to the point where it should be OK to
-        # terminate the connection.
-        #
-        # See http://www.snailbook.com/faq/background-jobs.auto.html for some more info.
-        _, raw_stderr = conn.communicate(
-            input=f"""
-          set -e
-          sudo mkdir -p {remote_dir}
-          {sshfs_command}
-          echo TRANSIENT_SSHFS_DONE
-          exit
-        """.encode(
-                "utf-8"
-            ),
-            timeout=sshfs_timeout,
-        )
 
-        # Ensure returncode is set
-        conn.poll()
+    # Everything from here on out simply verifies that nothing went wrong.
+    if ssh_proc.stderr is not None:
```

Comment:
Why would stderr ever be None? You setup the subprocess with PIPE

Reply:
It wouldn't, and this bugs me too. I had to add this if statement to appease mypy.